File: Experimental/2023/MattsToys/MattsToy2022-Beta1/src/main/cpp/commands/TankDrive.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/TankDrive.h"

TankDrive::TankDrive(DriveBase* driveBase,
                     std::function<double()> leftPowerSupplier,
                     std::function<double()> rightPowerSupplier)
    : m_driveBase(driveBase),
      m_leftPowerSupplier(leftPowerSupplier),
      m_rightPowerSupplier(rightPowerSupplier) {
  AddRequirements(driveBase);
}

// Called when the command is initially scheduled.
void TankDrive::Initialize() {
  m_driveBase->TankDrive(m_leftPowerSupplier(), m_rightPowerSupplier());
}

// Called repeatedly when this Command is scheduled to run
void TankDrive::Execute() {
  m_driveBase->TankDrive(m_leftPowerSupplier(), m_rightPowerSupplier());
}

// Called once the command ends or is interrupted.
void TankDrive::End(bool interrupted) {
  m_driveBase->Stop();
}

File: Experimental/2023/MattsToys/PidExamples/src/main/java/frc/robot/commands/TankDrive.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.robot.subsystems.DriveBase;

public class TankDrive extends CommandBase {
  public interface SpeedSupplier {
    double percent();
  }

  private final DriveBase m_driveBase;
  private final SpeedSupplier m_leftSupplier;
  private final SpeedSupplier m_rightSupplier;

  /** Creates a new TankDrive. */
  public TankDrive(DriveBase driveBase, SpeedSupplier leftSupplier, SpeedSupplier rightSupplier) {
    addRequirements(driveBase);
    m_driveBase = driveBase;
    m_leftSupplier = leftSupplier;
    m_rightSupplier = rightSupplier;
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
    m_driveBase.tankDrive(m_leftSupplier.percent(), m_rightSupplier.percent());
  }

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    m_driveBase.tankDrive(m_leftSupplier.percent(), m_rightSupplier.percent());
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_driveBase.stop();
  }
}
